### DISCUSSION TITLE
Option for Collection->shift() to always return a collection

### DISCUSSION THREAD

**Discussion by mjomble** [laravel/framework#51696] [Ideas]
The current behavior of Collection->shift() (since merging #51686) is:

- `shift(0)` returns a Collection
- `shift(1)` returns a single item ⚠️ 
- `shift(2)` returns a Collection
- `shift(3)` returns a Collection
- etc

This can easily lead to bugs (as just happened to me) if you pass a variable as the parameter, e.g. `shift($count)`.
A current workaround is to use `Collection::wrap($coll->shift($count))`, but it's not immediately obvious that such a wrapper exists, or that it's necessary to use it.

I propose adding an optional second boolean parameter to shift() which forces it to always return a Collection, even when shifting exactly one element. So that:

- `shift(0, true)` returns a Collection
- `shift(1, true)` returns a Collection
- `shift(2, true)` returns a Collection
- `shift(3, true)` returns a Collection
- etc

whereas `shift(1)` remains backwards-compatible.

**Comment by wizzymore** [comment]
But why?

It must return the item it removes AS IS, that's what shifting always did no matter the language or the object it was on, it returns the value it removes.

**Reply by mjomble** [reply]
This is a perfectly valid argument for the default case where shift() is invoked without any arguments. Then it returns a single item, exactly like PHP's array_shift() and the shift() method in various other languages and libraries.

This scenario would remain unaltered by this change. If invoked without any arguments, Collection->shift() would always return the first item as is.

However, Laravel's shift() differs from most other languages and libraries by introducing an additional $count parameter, which alters the behavior in cases where $count is not equal to 1. In those situations, the method returns a Collection, not a single item. The items are no longer returned as is, but wrapped in a Collection.
As far as I know, there is no common industry standard for such variation of the shift() method, it seems fairly Laravel-specific.

To illustrate the problem I'm trying to solve, consider an example like this:

```
function processItems(Collection $queue, int $count) {
    $items = $queue->shift($count);
    foreach ($items as $item) {
        // Do something with the item
    }
}
```

This code contains a bug that can be difficult to spot - if $count is 1, $items is not a Collection and you can't use it in foreach. I myself have created a bug like this, by assuming that shift() always returns a Collection if a $count argument is passed.

In hindsight, perhaps a cleaner and safer solution would've been to split this functionality between two separate methods, e.g:

- Collection->shift() which takes no arguments, always shifts exactly one item and returns it as is
- Collection->shiftMany($count) which always returns a Collection, even if $count = 1

This is no longer an option, since shift() has already been modified with the optional count parameter.

However, adding a shiftMany() method would still be a viable alternative to the additional parameter I've suggested above.

In any case, both approaches would keep `shift()` backwards compatible when invoked without any parameters.

**Reply by wizzymore** [reply]
I'm sorry man but adding a braking change or changing a function that works correctly for fixing a bug where the programmer is at fault is not a good motivator, atleast not for me.

Why not make a macro and do something like this and always wrap the answer and boom, problem fixed:
```php
Collection::macro('shiftWrap', function (int $count = 1): Collection {
    return self::wrap($this->shift($count));
});
```

And call it like $queue->shiftWrap($count) and it always returns a collection.

The idea of this Macroable function is that when something doesn't act like you want it to, you can make it to.

**Reply by mjomble** [reply]
As far as I can tell, I am not suggesting a breaking change.
Everything would be fully backwards compatible.

If I am wrong, could you point out which part of my suggestion constitutes a breaking change?

**Reply by henzeb** [reply]
It is bullshit to have one extra parameter when we could easily check if one was passed in... The boolean would only do something when the value is 1.

Yes, that would be a breaking change, but one that makes sense ...

**Reply by mjomble** [reply]
@henzeb are you saying my recommendation is "bullshit" or am I misreading your comment?

**Reply by henzeb** [reply]
The idea is good, just the extra parameter is bullshit. This must be a breaking change for laravel 12. A workaround until the might be a macro.

**Reply by mjomble** [reply]
> The idea of this Macroable function is that when something doesn't act like you want it to, you can make it to.

The goal of this change is not to fix the bug in my project only. I have already done that and know to watch out for this from now on.
However, it seems likely to me that other developers may make the same assumption and end up with similar bugs.
The reason I started this discussion was to try to reduce the risk of that happening.

> This must be a breaking change for laravel 12

I'm not sure if a breaking change is needed here. But I'm also not very familiar with Laravel's policy on breaking changes.

As a non-breaking alternative, what about keeping shift() as is and adding a new shiftMany() method that would always return a Collection?

PS. Laravel has a [code of conduct](https://laravel.com/docs/11.x/contributions#code-of-conduct) which states, for example:
> Participants must ensure that their language and actions are free of personal attacks and disparaging personal remarks

This is my first attempt at contributing an improvement to this project.
So far I've been told that the change, which I carefully designed to be backwards-compatible, is breaking (with a hint that my bug was 100% my fault and had nothing to do with the framework behaving surprisingly) and that a part of my proposal is bullshit.

I don't consider myself an easily-offended snowflake who can't handle any negative words. However, my free time is quite limited and I'm starting to feel like it would be better used elsewhere.

**Reply by henzeb** [reply]
well, I am not going into that kind of discussion again...  the word bullshit is in no way an attack to you. 

I agree with your idea that this has to be fixed, I just don't think we need a "backwards compatible" change here. 

This is my final message in this thread.

**Reply by mjomble** [reply]
> well, I am not going into that kind of discussion again... the word bullshit is in no way an attack to you.

Thanks for clarifying your intent 👍

I agree that it's pointless to argue over whether or not it is objectively an attack.
I also hope we can agree that even if you don't consider it an attack, some people may still subjectively interpret it as one (seems like this was not the first time it happened).

If you keep telling people their ideas are bullshit, some may interpret your intent correctly while others may not.
Whether you want to change your wording to avoid that is up to you.
